_posts/2023-02-22-reveal-type-in-crystal.md

```diff
@@ -0,0 +1,397 @@
+end
+
+macro reveal_type(t)
+  {% loc = "#{t.filename.id}:#{t.line_number}:#{t.column_number}" %}
+  {% puts "Revealed type #{loc.id}" %}
+  {% puts "  #{t.id}" %}
+  reveal_type_helper({{t}})
+end
+{% endraw %}
+```
+
+With this snippet at the top of our `program.cr` we already have the output we want. 🎉
+
+```console
+% crystal build program.cr
+Revealed type /path/to/program.cr:14:15
+  x
+  : (String | Nil)
+```
+
+Unfortunately if we put multiple `reveal_type` invocations things will not work as expected. The macro at `reveal_type_helper` is executed only **once per distinct type**.
+
+To force a different `reveal_type_helper` instance for each `reveal_type` invocation we would need a distinct type for each. Surprisingly we can do that.
+
+```crystal
+{% raw %}
+def reveal_type_helper(t : T, l) : T forall T
+  {% puts "   : #{T}" %}
+  t
+end
+
+macro reveal_type(t)
+  {% loc = "#{t.filename.id}:#{t.line_number}:#{t.column_number}" %}
+  {% puts "Revealed type #{loc.id}" %}
+  {% puts "  #{t.id}" %}
+  reveal_type_helper({{t}}, { {{loc.tr("/:.","___").id}}: 1 })
+end
+{% endraw %}
+```
+
+The `l` argument will have a tuple of type `{ <loc>: Int32 }` where `<loc>` is an identifier that depends on the invocation location of the `reveal_type` macro. 🤯
+
+
+## Caveats
+
+There are a couple of more caveats worth mentioning of this solution. A proper built-in feature in the compiler would not be affected by all of them. Essentially these can be sum up as:
+
+
+- The `reveal_type` needs to be within used code
+- Our implementation is very sensible to internals compiler’s execution order
+- It doesn't handle fully recursive definitions
+- It could change the semantic of the program since it affects the memory layout
+
+Feel free to skip to jump to the [next section](#ideas-for-the-compiler) unless you want some further details and examples of each.
+
+Due to how Crystal compiler works, the `reveal_type` needs to appear within **static reachable code**. Even if you start in a `def` with arguments with types (type restrictions actually) you need that `def` to be called. Otherwise the compiler ignores it. Similar to how C++ templates are not expanded unless they are used.
+
+Splitting the desired output between macro and defs is very sensible to the **compiler’s execution order**. The following code suffers from this:
+
+```crystal
+"a".tap do |a|
+  reveal_type a
+end
+
+1.tap do |a|
+  reveal_type a
+end
+```
+
+```console
+Revealed type /path/to/program.cr:2:15
+  a
+Revealed type /path/to/program.cr:6:15
+  a
+    : String
+    : Int32
+```
+
+**Recursive** programs can hit an edge case that would hide the output of `reveal_type_helper`. The following program will have several `dig_first` instantiation. As such the `reveal_type` macro will be called once per each, but all of the `reveal_type_helper` invocations are with the same `t` type and in the same location. We fall again in the problem we solved earlier with the `{ <loc>: Int32 }` param.
```

Comment:
```suggestion
**Recursive** programs can hit an edge case that would hide the output of `reveal_type_helper`. The following program will have several `dig_first` instantiation. As such the `reveal_type` macro will be called once per each (should there be another word here, each <something>, or macro can be called once each time?), but all of the `reveal_type_helper` invocations are with the same `t` type and in the same location. We fall again in the problem we solved earlier with the `{ <loc>: Int32 }` param.
```

Reply:
I think the omission refers to "instantiation". Wouldn't hurt to spell it out in order to avoid confusion.